akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -2285,6 +2285,281 @@ private[akka] final class StatefulMap[S, In, Out](
+      }
+
+      private def releaseThenCompleteStage(): Unit = {
+        if (stateAcquired) {
+          state.foreach(s => {
+            try {
+              val future = onComplete(s)
+              stateAcquired = false
+              future.value match {
+                case None =>
+                  future.onComplete(getAsyncCallback[Try[Option[Out]]] {
+                    case Failure(e) => failStage(e)
+                    case Success(output) =>
+                      output match {
+                        case Some(v) => emit(out, v, () => initState())
+                        case None    => initState()
+                      }
+                  }.invoke)(akka.dispatch.ExecutionContexts.parasitic)
+                case Some(v) =>
+                  emitThenFailOrCompleteStage(None, v)
+              }
+            } catch {
+              // ideally should never throw when release resource
+              case NonFatal(e) => failStage(e)
+            }
+          })
+        }
+      }
+
+      override def onUpstreamFinish(): Unit =
+        if (buffer.isEmpty && earlyPulledElem.isEmpty) {
+          releaseThenCompleteStage()
+        }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = if (state.isDefined) {
+        releaseThenFailStage(ex)
+      }
+
+      override def onDownstreamFinish(cause: Throwable): Unit = if (state.isDefined) {
+        releaseThenFailStage(cause)
+      }
+
+      override def postStop(): Unit = {
+        state match {
+          case Some(value) =>
+            // outlet isn't available here
+            if (stateAcquired) onComplete(value)
+          //TODO figure out what should be done here
+          // since resource maybe in the half way of acquisition
+          case None =>
+        }
+      }
+
+      override def preStart(): Unit = {
+        initState()
+        buffer = BufferImpl(parallelism, inheritedAttributes)
+      }
+
+      override def onPull(): Unit = pushNextIfPossible()
+
+      override def onPush(): Unit = {
+        val elem = grab(in)
+        state match {
+          case Some(_) =>
+            applyUserFunction(elem)
+            pullIfNeeded()
+          case None =>
+            earlyPulledElem = Some(elem)
+            pushNextIfPossible()
+        }
+      }
+
+      @tailrec
+      private def pushNextIfPossible(): Unit =
+        if (state.isEmpty) tryPushAfterInitialized = true
+        else if (earlyPulledElem.isDefined) applyUserFunction(earlyPulledElem.get)
+        else if (buffer.isEmpty) pullIfNeeded()
+        else if (buffer.peek().elem eq NotYetThere) pullIfNeeded() // ahead of line blocking to keep order
+        else if (isAvailable(out)) {
+          val holder = buffer.dequeue()
+          holder.elem match {
+            case Success(elem) =>
+              if (elem != null && elem._1 != null && elem._2 != null) {
+                // state has to be presented here
+                state = Some(combineState(state.get, elem._1))
```

Comment:
combineState is only used here. `state.get` can be state match ?

Reply:
Already checking if `state` is existed before